gen/app-gen

```diff
@@ -240,6 +247,136 @@ export_image() {
     fi
 }
 
+function parse_parent_child() {
+    local -r root_dir="$1"/parent-child
+    local -A id2parent
+    local -A parent2id
+    local -a ids
+    local -a a
+    local json
+    local i
+    local parentid
+    local id
+
+    while read -r json; do
+        parentid=$(cat "$json" | jq -r .parent)
+        id=$(dirname "$json")
+        id=$(basename "$id")
+        id2parent[${id}]=${parentid}
+        parent2id[${parentid}]=${id}
+        [[ "$parentid" != "null" ]] && ids+=("${id}")
+    done < <(find "$1" -name json)
+
+    set +e
+    parentid="${parent2id[null]}"
+    mkdir -p "${root_dir}/${parentid}"
+    i=0
+    while [[ ${#ids[@]} -gt 0 ]]; do
+        id=${ids[${i}]}
+        parentid=${id2parent[${id}]}
+        [[ "$parentid" == "" ]] && { break; }
+        parent_dir=$(find "$root_dir" -name "$parentid" -and -type d)
+        if [[ "$parent_dir" != "" ]]; then
+            mkdir -p "${parent_dir}/${id}"
+            a=("${ids[@]/${id}/}")
+            ids=(${a[@]})
+            i=0
+        else
+            let i++
+            [[ $i -ge ${#ids[@]} ]] && i=0
+        fi
+    done
+    set +e
```

Comment:
:question:  why `set +e` both places?

Reply:
good catch, that should have been `set -e` here. thanks.